core/src/main/java/net/adoptopenjdk/icedteaweb/resources/downloader/PrioritizedParallelExecutor.java

```diff
@@ -0,0 +1,220 @@
+        private final CompletableFuture<V> lowerPriorityResult = new CompletableFuture<>();
+        private final AtomicReference<Future<V>> lowerPriority = new AtomicReference<>();
+
+        private final Future<V> delegate;
+
+        private ChainableFuture(Future<V> delegate) {
+            this.delegate = delegate;
+        }
+
+        /**
+         * Set a lower priority to be canceled.
+         * This is not passed in the constructor as the value is not known then.
+         *
+         * @param lowerPriority the lower priority future to be canceled this future is canceled.
+         */
+        void setLowerPriority(Future<V> lowerPriority) {
+            this.lowerPriority.set(lowerPriority);
+
+            if (terminatedExceptionally.get()) {
+                Executors.newSingleThreadExecutor().execute(() -> {
+                            try {
+                                lowerPriorityResult.complete(lowerPriority.get());
+                            } catch (Exception e) {
+                                lowerPriorityResult.completeExceptionally(e);
+                            }
+                        }
+                );
+            }
+        }
+
+        @Override
+        public boolean cancel(boolean mayInterruptIfRunning) {
+            final boolean result = delegate.cancel(mayInterruptIfRunning);
+            cancelLowerPriority();
+            return result;
+        }
+
+        @Override
+        public boolean isCancelled() {
+            return delegate.isCancelled();
+        }
+
+        @Override
+        public boolean isDone() {
+            return delegate.isDone();
+        }
+
+        @Override
+        public V get() throws InterruptedException, ExecutionException {
+            try {
+                return delegate.get();
+            } catch (ExecutionException e) {
+                return returnLowerPriority();
+            }
+        }
+
+        @Override
+        public V get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
+            final long start = System.nanoTime();
+            try {
+                return delegate.get(timeout, unit);
+            } catch (ExecutionException e) {
+                final long elapsed = System.nanoTime() - start;
+                final long remaining = unit.toNanos(timeout) - elapsed;
+                return returnLowerPriority(remaining, NANOSECONDS);
+            } catch (TimeoutException e) {
+                return returnLowerPriority(1, MILLISECONDS);
+            }
+        }
+
+        private void cancelLowerPriority() {
+            terminatedExceptionally.set(true);
+            final Future<V> lowerPriority = this.lowerPriority.get();
+            if (lowerPriority != null) {
+                lowerPriority.cancel(true);
+            }
+        }
+
+        private V returnLowerPriority() throws InterruptedException, ExecutionException {
+            terminatedExceptionally.set(true);
```

Comment:
The name "terminatedExceptionally" does not really fit to the usage since you mutate the field only in methods that are related to the lowerPrioFuture

Reply:
moved setters to make name more fitting